### ISSUE TITLE
Why and when `app.base.cache.data.name` is overriden by generate ?

### ISSUE THREAD

**Issue by rbecheras** [generate/generate#50]
Why and when `app.base.cache.data.name` and `app.base.data(name)`  are overriden by generate ?

I'm developing a generator that: 

- use `app.ask()` to ask for data but not for `name`
- set manually `app.base.data({name})` in function of other user answers once it get these answers
- use `<%= name >` in some templates to be rendered

I don't understand why after some template rendering (performed in some dedicated tasks), `app.base.cache.data.name` is reset to the `cwd`, overriding the manual value it put onto.

The consequence is some templates are rendered with the good value, whereas some other are filled with the `cwd`.

**But, if FIRST I ask for `name`**, i.e.:

```js
app.question('name', opts)
app.ask(['name'], ({name}) => app.base.data({name}))`
```
**Then  `app.base.cache.data.name`  neither app.base.data(name) is not overridden**

Do you see what I mean?  and why and how it behaves like that ?

More generally, I have some difficulties to understand the what, why and how of:

- `app.data`
- `app.base.data`
- `app.data.cache`
- `app.base.data.cache`
- `app.cache`
- `app.cache.data`

And their interactions with `app.ask()` or `app.question()` and `app.choices()...`

Can you brief me about that please ?

**Comment by jonschlinkert** [comment]
```js
app.data
app.base.data
app.data.cache
app.base.data.cache
app.cache
app.cache.data
```

TLDR; there are really only two properties:

- `app.cache` - an object
- `app.data` - a method 

and

- [app.base](https://github.com/node-base/base/blob/master/index.js#L198) - which is used for accessing the "base" instance of `app` when instances are nested (more detail and examples below). Since `app.base` is just a "parent" instance of `app`, both the `.cache` and `.data` properties will be found on that instance. 

**app.cache**

- `app.cache` is an object that is used for storing user-defined values. this allows us to keep those values off of the root of the `app` object. 
- `app.data()` is a method for getting and setting values on the `app.cache.data` object. The `data` object is the only property we reserve on `app.cache`... I think. We use this object for caching values that will be used for resolving templates.

**app.data.cache**

This is a private object used internally by the `app.data()` method. I thought this object was non-enumerable, if it's not it should be. Did this show up in your data?

**app.base**

`app.base` is the "base" instance when nested apps are registered. The following examples show how this works, it will make more sense afterwards. 

First, when you register a generator, a new instance of Generate is created for that generator:

```js
const Generate = require('generate');
const app = new Generate({ silent: true });

app.register('foo', function(foo) {
   // "foo", a new instance of Generate, is cached on the `app.generators` object 
});

console.log(app.generators);
```

Since a generator is just an instance of Generate, there is nothing stopping generators from registering other generators. Which means they can be "nested" any number of levels deep. For example:

```js
const Generate = require('generate');
const app = new Generate({ silent: true });

// "app" is the base instance
app.register('foo', function(foo) {
  console.log(foo.base === app); //=> true

  foo.register('bar', function(bar) {
    console.log(bar.base === app); //=> true

    bar.register('baz', function(baz) {
      baz.task('default', function(cb) {
        console.log(baz.base === app); //=> true
        cb();
      });
    });
  });
});

// log out the instances to see the relationships of the instances
console.log(app.generators);
console.log(app.generators.foo.generators);
console.log(app.generators.foo.generators.bar.generators);
```

At any level of nesting, we can always get the "base" instance (the first `app` we defined) using the `.base` property.  

**Fun fact**

Did you know you can run tasks on nested generators like this?

```js
app.generate('foo.bar.baz:default', function(err) {
  if (err) return console.error(err);
  console.log('Done!');
});
```

**Bonus**

The following code is a recreation of the relevant parts of [base](https://github.com/node-base/base) and [generate](https://github.com/generate/generate) to illustrate how the `app.base`, `app.cache` and `app.data` properties work in the simplest way possible:

```js
class Base {
  constructor(name) {
    this.name = name;
    this.generators = {};
    this.cache = {};
  }

  data(key, value) {
    if (arguments.length === 1) {
      return this.cache[key];
    }
    this.cache[key] = value;
    return this;
  }

  register(key, fn) {
    this.generators[key] = () => {
      const app = new this.constructor(key);
      app.parent = this; // we need to set the "parent" in order for `app.base` to work
      fn.call(app, app);
      return app;
    };
    return this;
  }

  get base() {
    return this.parent ? this.parent.base : this;
  }
}
```

Now, let's create an example application with the code, to see how the nesting works:

```js
const app = new Base('grandparent');
app.data('admin', true);

console.log(app.name);
console.log(app.parent.name); // app.parent is undefined
console.log(app.base.name);
console.log()

app.register('foo', function() {
  console.log(this.name);
  console.log(this.parent.name);
  console.log(this.base.name);
  console.log()

  this.register('bar', function() {
    console.log(this.name);
    console.log(this.parent.name);
    console.log(this.base.name);
    console.log()

    this.register('baz', function() {
      console.log(this.name);
      console.log(this.parent.name);
      console.log(this.base.name);
      console.log()

      this.register('qux', function() {
        console.log(this.name);
        console.log(this.parent.name);
        console.log(this.base.name);
        console.log()
      });
    });
  });
});

app
  .generators.foo()
  .generators.bar()
  .generators.baz()
  .generators.qux()
```

Note that `app.base.name` is always the same, no matter what instance you're on.

**Comment by jonschlinkert** [comment]
btw, sorry for the delay in answering. I've been out of the office a lot the past week

**Comment by rbecheras** [comment]
Ok don't worry for the delay, I know you have lot of work, and I see you answer soon or late.

Ok I understand all of your explanations.

But so, when I want to set a given value with: 

```js
app.base.data('name', myValue)
```

in a sub-generator, why that value is reset some time later during the generation of a following task that render a template ?

But if instead of just doing `app.base.data('name', myValue)`, I do:

```js
app.question('name', {...})
app.ask(['name'], ({name}) => app.base.data('name', name))
```

Why in that case, the value is not reset at any time ?
